src/nsrt_learning/option_learning.py

```diff
@@ -636,16 +685,215 @@ class _DummyRLOptionLearner(_RLOptionLearnerBase):
+        self.min_sigma = min_sigma
+        self.decay_period = decay_period
+        self.action_dim = action_space.shape[0]
+        self.low = action_space.low
+        self.high = action_space.high
+        self.reset()
+
+    def reset(self) -> None:
+        self.state = np.ones(self.action_dim) * self.mu
+
+    def evolve_state(self) -> Array:
+        x = self.state
+        dx = self.theta * (self.mu - x) + self.sigma * np.random.randn(
+            self.action_dim)
+        self.state = x + dx
+        return self.state
+
+    def get_action(self, action: Array, t: int = 0) -> Array:
+        ou_state = self.evolve_state()
+        self.sigma = self.max_sigma - (self.max_sigma - self.min_sigma) * min(
+            1.0, t / self.decay_period)
+        return np.clip(action + ou_state, self.low, self.high)
+
+
+class _DDPGAgent(_RLOptionLearnerBase):
+    """Updates the policy via Deep Deterministic Policy Gradient
+    (https://arxiv.org/abs/1509.02971).
+
+    Note that DDPG is not executed exactly: "episodes" in the algorithm
+    correspond to "online learning cycles" in our code, and we update
+    the critic and actor after a large set of interactions, not after
+    each interaction with the environment.
+    """
+
+    def __init__(self) -> None:
+        self._initialized = False
+        self._seed = CFG.seed
+        self._batch_size = CFG.nsrt_rl_batch_size
+        self._tau = CFG.nsrt_rl_tau  # polyak averaging parameter
+        self._gamma = CFG.nsrt_rl_discount_factor  # RL discount factor
+        self._max_size = CFG.nsrt_rl_replay_buffer_size
+
+    def update(
+        self, option: _LearnedNeuralParameterizedOption,
+        experience: List[List[Tuple[State, Array, Action, int, State, Array,
+                                    bool]]]
+    ) -> Union[_NoisyNeuralParameterizedOption,
+               _LearnedNeuralParameterizedOption]:
+        # If not initialized, create replay buffer and actor-critic networks.
+        if not self._initialized:
+            # Initialize replay buffer.
+            example_option_run = experience[0]
+            example_transition = example_option_run[0]
+            state_dim = example_transition[1].shape[0]
+            action_dim = example_transition[2].arr.shape[0]
+            self._replay_buffer = ReplayBuffer(CFG.seed, self._max_size,
+                                               state_dim, action_dim)
+
+            # Note that the actor has _x_dim and _y_dim set
+            # already.
+            self._actor = cast(MLPRegressor, option.regressor)
+            self._target_actor = copy.deepcopy(self._actor)
+            input_dim = self._actor._x_dim
+            output_dim = self._actor._y_dim
+            self._critic = Critic(CFG.nsrt_rl_critic_regressor_hid_sizes,
+                                  state_dim + action_dim)
+            self._critic_loss_fn = nn.MSELoss()
+            self._target_critic = copy.deepcopy(self._critic)
```

Comment:
deepcopy is generally slow and discouraged, but maybe this is the way that other implementations do this?